File: lib/js/phantom.dart
import 'dart:typed_data';

import 'package:js/js.dart';
import 'package:sol_dapp_flutter_test/js/solana_web3/transaction.dart';
import 'package:sol_dapp_flutter_test/js/solana_web3/public_key.dart';

@JS('wallet.phantom.connect')
external Future<PublicKey> connect();

@JS('wallet.phantom.disconnect')
external Future<PublicKey> disconnect();

@JS('wallet.phantom.isPhantomInstalled')
external bool? isPhantomInstalled();

@JS('wallet.phantom.signTransaction')
external Future<String> signTransaction(Transaction tx);

@anonymous
@JS('wallet.phantom.SignedMessage')
class SignedMessage {
  external factory SignedMessage({Uint8List signature, PublicKey publicKey});
  external Uint8List get signature;
  external PublicKey get publicKey;
}

@JS('wallet.phantom.signMessage')
external Future<SignedMessage> signMessage(String msg);

File: src/phantom.ts
const solanaWeb3 = require('@solana/web3.js');

export async function connect(): solanaWeb3.PublicKey {
  const resp = await window.solana.connect();
  return new solanaWeb3.PublicKey(resp.publicKey.toString());
}

export function disconnect() {
  window.solana.disconnect();
}

export function isPhantomInstalled(): bool {
  return window.solana && window.solana.isPhantom;
}

export async function signTransaction(tx: solanaWeb3.Transaction): solanaWeb3.Transaction {
  return await window.solana.signTransaction(tx);
}

type SignedMessage = {
    signature: string,
    publicKey: solanaWeb3.PublicKey
}

export async function signMessage(msg: string): SignedMessage {
  const encodedMsg = new TextEncoder().encode(msg);
  return await window.solana.signMessage(encodedMsg, 'utf8');
}
